R/conformal_infer.R

```diff
@@ -0,0 +1,459 @@
+  }
+
+  # Compute the abs residuals then more statistics
+  tmp_res <-
+    augment(tmp_fit, trial_data) |>
+    dplyr::mutate(
+      .abs_resid = abs(!!rlang::sym(y_name) - .pred)
+    )
+  # Quantiles of original training set (abs) residuals
+  quant_val <- stats::quantile(tmp_res$.abs_resid[-1], prob = level)
+
+  res <-
+    dplyr::tibble(
+      quantile = unname(quant_val),
+      trial = trial,
+      .abs_resid = tmp_res$.abs_resid[nrow(trial_data)]
+    )
+  res$difference <- res$.abs_resid - res$quantile
+  res
+}
+
+# ------------------------------------------------------------------------------
+
+grid_all <- function(new_data, model, train_data, level, ctrl) {
+
+  furrr::future_map_dfr(
+    new_data,
+    grid_one,
+    model,
+    train_data,
+    level,
+    ctrl,
+    .progress = ctrl$progress,
+    .options = furrr::furrr_options(seed = ctrl$seed, stdout = FALSE)
+  )
+}
+
+grid_one <- function(new_data, model, train_data, level, ctrl) {
+  y_name <- get_outcome_name(model)
+
+  pkg_inst <- lapply(ctrl$required_pkgs, rlang::is_installed)
+  pred_val <- new_data$.pred
+  bound <- new_data$.bound
+  new_data <- dplyr::select(new_data, -.pred, -.bound)
+  new_data[[y_name]] <- NA_real_
+  trial_data <- dplyr::bind_rows(train_data, new_data)
+
+  trial_vals <- seq(pred_val - bound, pred_val + bound, length.out = ctrl$trial_points)
+
+  res <-
+    purrr::map_dfr(
+      trial_vals,
+      ~ trial_fit(.x,
+                  trial_data = trial_data,
+                  wflow = model,
+                  level = level)
+    )
+
+  compute_bound(res, pred_val)
+}
+
+compute_bound <- function(x, predicted) {
+  x <- x[stats::complete.cases(x),]
+  if (all(x$difference < 0)) {
+    rlang::warn("Could not determine bounds.")
+    res <-
+      dplyr::tibble(.pred_lower = NA_real_,
+                    .pred = predicted,
+                    .pred_upper = NA_real_)
+    return(res)
+  }
+
+  upper <- x[x$trial >= predicted & x$difference >= 0,]
+  if (nrow(upper) > 0) {
+    upper <- min(upper$trial)
+  } else {
+    upper <- NA_real_
+  }
+  lower <- x[x$trial <= predicted & x$difference >= 0,]
+  if (nrow(lower) > 0) {
+    lower <- max(lower$trial)
+  } else {
+    lower <- NA_real_
+  }
+  dplyr::tibble(.pred_lower = lower, .pred_upper = upper)
+}
+
+# ------------------------------------------------------------------------------
+
+optimize_all <- function(new_data, model, train_data, level, ctrl) {
+  # purrr::map_dfr(
+  furrr::future_map_dfr(
+    new_data,
+    optimize_one,
+    model = model,
+    train_data = train_data,
+    level = level,
+    ctrl = ctrl,
+    .progress = ctrl$progress,
+    .options = furrr::furrr_options(seed = ctrl$seed, stdout = FALSE)
+  )
+}
+
+get_diff <- function(val, ...) {
+  res <- trial_fit(val, ...)
+  res$difference
+}
+
+optimize_one <- function(new_data, model, train_data, level, ctrl) {
+  y_name <- get_outcome_name(model)
+
+  pkg_inst <- lapply(ctrl$required_pkgs, rlang::is_installed)
```

Comment:
```suggestion
  pkg_inst <- lapply(ctrl$required_pkgs, rlang::check_installed)
```

Same note on `rlang::is_installed()` vs. `rlang::check_installed()`.